File: julia/diag_gsic_blocks_vs_emulandice.jl
## ============================================================================
## diag_gsic_blocks_vs_emulandice.jl — Ladrillo's three glacier reservoirs on the
## SHARED FACTS climate, sample-matched to emulandice.
##
## THE QUESTION (Marcus 2026-09-14). Ladrillo projects ~25% less glacier melt at 2100
## than emulandice (the GlacierMIP2-trained GP that FACTS runs) on every scenario. Is
## that one block or all three, and is it the driver or the module? emulandice is
## GSAT-driven and ran on facts/experiments/global.shared.<scen>.n200 — 200 configs of
## the same FaIR cube, same 2014 splice, same 1995-2014 reference Ladrillo uses. So
## sample k of emulandice and sample k here saw the SAME GSAT path, and any residual
## is the glacier model (law, commitment, response time, regional amplification,
## initial disequilibrium), not the climate.
##
## ARMS.
##   joint  : posterior draw k on shared sample k (climate + parameter spread) — the
##            like-for-like arm against emulandice's climate + GP sampling.
##   fixed  : every draw on the MEAN of the 200 shared paths (parameter spread only),
##            so joint-minus-fixed is the climate's share of the spread.
##   joint_cmip6amp : joint with the per-block amplification overridden by the CMIP6
##            regional-characteristic ratio (see the arm's comment) — the driver lever.
## Draws: NDRAW rows of the canonical L24 subsample chosen by a SEEDED permutation;
## the seed and the row index of every draw are written into the output.
##
## OUTPUT (cm SLE, cumulative melt since the model start, i.e. NOT re-referenced;
## the python side rebases both models to FACTS's base year 2005):
##   outputs/diag_emu_blocks/ladrillo_blocks_<scen>.csv
##   columns scen, arm, sample, draw_row, block (R19/SLOWP/FAST), year, melt_cm
##   + ladrillo_blocks_<scen>_provenance.txt (driver, posterior, seed, climate, units)
##
##   julia --project=julia_v2 julia/diag_gsic_blocks_vs_emulandice.jl --scen=ssp245 [--ndraw=200] [--seed=2026]
## Needs outputs/diag_emu_blocks/shared_<scen>_{gmst,ohc}.csv from
## python/dump_shared_climate_for_blocks.py.
## ============================================================================
using CSV, DataFrames, Mimi, Printf, Statistics, Random
include(joinpath(@__DIR__, "ladrillo_projection.jl"))

_argval(p) = (i = findfirst(a -> startswith(a, p), ARGS); i === nothing ? nothing : ARGS[i][length(p)+1:end])
const SCEN  = something(_argval("--scen="), "ssp245")
const NDRAW = parse(Int, something(_argval("--ndraw="), "200"))
const SEED  = parse(Int, something(_argval("--seed="), "2026"))
const DIR   = joinpath(LADRILLO_REPO, "outputs/diag_emu_blocks")
const OUT   = joinpath(DIR, "ladrillo_blocks_$(SCEN).csv")
const YEARS_OUT = vcat(collect(2005:5:2100), [2150, 2300])
const SLOTS = ("R19" => :gsic_r19, "SLOWP" => :gsic_slowp, "FAST" => :gsic_fast)
const Y0, Y1 = 1850, 2300

cg = CSV.read(joinpath(DIR, "shared_$(SCEN)_gmst.csv"), DataFrame)
co = CSV.read(joinpath(DIR, "shared_$(SCEN)_ohc.csv"), DataFrame)
@assert Int.(cg.year) == collect(Y0:Y1) "gmst dump is not $(Y0):$(Y1)"
@assert names(cg) == names(co) "gmst/ohc dumps disagree on samples"
const SAMPLES = [c for c in names(cg) if startswith(c, "s")]
const NS = length(SAMPLES)
NS >= NDRAW || error("only $(NS) shared samples, asked for $(NDRAW) draws")
gpath(k) = Float64.(cg[!, SAMPLES[k]])
opath(k) = Float64.(co[!, SAMPLES[k]])
const MEAN_G = vec(mean(Matrix(cg[!, SAMPLES]), dims=2))
const MEAN_O = vec(mean(Matrix(co[!, SAMPLES]), dims=2))

## ---- draws: seeded permutation of the canonical subsample ------------------
post = ladrillo_posterior()
const DRAW_ROWS = randperm(MersenneTwister(SEED), nrow(post))[1:NDRAW]
const VARIANT = ladrillo_posterior_variant()
const PROV = "driver=diag_gsic_blocks_vs_emulandice.jl; posterior=$(basename(LADRILLO_POSTERIOR_CSV)) " *
             "($(nrow(post)) rows); draws=$(NDRAW) by randperm(MersenneTwister($(SEED))) applied " *
             "immediately before the draw selection; climate=shared_$(SCEN) (FACTS global.shared." *
             "$(SCEN).n200 input, spliced@2014, ref 1995-2014, K rel 1850-1900); run $(Y0)-$(Y1); " *
             "units cm SLE cumulative melt since $(Y0) (no re-referencing); nu basis $(LADRILLO_NU_BASIS)"

@printf("Ladrillo glacier blocks on the shared FACTS climate | %s | %d samples, %d draws (seed %d)\n",
        SCEN, NS, NDRAW, SEED)
@printf("  GMST@2100 over the samples: median %.2f [%.2f, %.2f] K; mean path %.2f\n",
        median([gpath(k)[end-200] for k in 1:NS]), quantile([gpath(k)[end-200] for k in 1:NS], 0.05),
        quantile([gpath(k)[end-200] for k in 1:NS], 0.95), MEAN_G[end-200])
flush(stdout)

rows = DataFrame(scen=String[], arm=String[], sample=Int[], draw_row=Int[], block=String[],
                 year=Int[], melt_cm=Float64[])

function record!(bf, arm, k, drow)
    for (b, slot) in SLOTS
        s = Float64.(bf.m[:glaciers_small_icecaps, slot])   # m SLE since 1850, raw
        for y in YEARS_OUT
            push!(rows, (SCEN, arm, k, drow, b, y, 100.0 * s[ladrillo_yi(bf, y)]))
        end
    end
end

## arm `fixed`: every draw on the ensemble-mean shared path
let bf = ladrillo_setup(ssp="ssp245", y0=Y0, y1=Y1, gis_variant=VARIANT, gmst=MEAN_G, ohc=MEAN_O)
    @printf("  arm fixed ... "); flush(stdout)
    for k in 1:NDRAW
        ladrillo_run_draw!(bf, post[DRAW_ROWS[k], :])
        record!(bf, "fixed", k, DRAW_ROWS[k])
    end
    println("done")
end
## arm `joint`: draw k on shared sample k (one model build per sample)
## arm `joint_cmip6amp`: the same, with every draw's per-block amplification REPLACED by
## the CMIP6 regional-characteristic ratio (`amp_regchar` in extc_block_constants.csv:
## the GCM-pattern value a GSAT-driven emulator trained on GCM-forced runs implicitly
## carries). Ladrillo's fitted amps are R19 0.72 / SLOWP 2.50 / FAST 1.45 (prior centres;
## sampled per draw); the CMIP6 values are 1.03 / 1.70 / 1.23. The arm measures how much
## of any block gap is the DRIVER amplification rather than the glacier law. Projection-
## side override only (the posterior was fitted with its own amps) — a lever, not a refit.
const BC = CSV.read(LADRILLO_BLOCK_CONSTANTS_CSV, DataFrame)
const AMP_CMIP6 = Dict(String(r.block) => Float64(r.amp_regchar) for r in eachrow(BC))
@printf("  CMIP6 regional-characteristic amps: %s\n",
        join(["$(b)=$(round(AMP_CMIP6[b], digits=2))" for b in LADRILLO_BLOCKS], " "))
for (arm, override) in (("joint", false), ("joint_cmip6amp", true))
    @printf("  arm %s ... ", arm); flush(stdout)
    for k in 1:NDRAW
        bf = ladrillo_setup(ssp="ssp245", y0=Y0, y1=Y1, gis_variant=VARIANT, gmst=gpath(k), ohc=opath(k))
        one = post[DRAW_ROWS[k]:DRAW_ROWS[k], :]        # one-row copy, mutable
        if override
            for b in LADRILLO_BLOCKS; one[1, "gic_amp_$b"] = AMP_CMIP6[b]; end
        end
        ladrillo_run_draw!(bf, one[1, :])
        record!(bf, arm, k, DRAW_ROWS[k])
        k % 50 == 0 && (@printf("%d ", k); flush(stdout))
    end
    println("done")
end

mkpath(DIR)
CSV.write(OUT, rows)
## provenance ONCE per file (a per-row column made the file 17 MB; the sidecar keeps the stamp)
open(replace(OUT, ".csv" => "_provenance.txt"), "w") do io; println(io, PROV); end
## summary: 2100 melt since 2005 per block, both arms
@printf("\n  %-6s %-6s %10s %18s\n", "arm", "block", "2100-2005", "5-95%")
for arm in ("fixed", "joint", "joint_cmip6amp"), (b, _) in SLOTS
    v = [rows.melt_cm[(rows.arm .== arm) .& (rows.block .== b) .& (rows.year .== 2100) .& (rows.sample .== k)][1] -
         rows.melt_cm[(rows.arm .== arm) .& (rows.block .== b) .& (rows.year .== 2005) .& (rows.sample .== k)][1]
         for k in 1:NDRAW]
    @printf("  %-6s %-6s %10.2f  [%6.2f, %6.2f]\n", arm, b, median(v), quantile(v, 0.05), quantile(v, 0.95))
end
println("\nwrote $(relpath(OUT, LADRILLO_REPO))")

File: python/diag_gsic_blocks_vs_emulandice.py
#!/usr/bin/env python3
"""diag_gsic_blocks_vs_emulandice.py — Ladrillo L24 glacier blocks vs FACTS emulandice, per
block, on the SAME 200 FaIR configs (sample-matched), 2005-2100.

WHAT IS COMPARED. emulandice (Edwards et al. 2021: GP emulators of the GlacierMIP2
transient runs, one per RGI region, GSAT-driven, 2015-2100) as FACTS ran it on the shared
climate (facts/experiments/global.shared.<scen>.n200, per-region outputs glac1..glac19,
decadal 2020-2100, base year 2005). Regions are summed to Ladrillo's blocks:
    SLOWG (code SLOWP) = RGI 03, 09, 07, 06
    FASTG (code FAST)  = the 13 others except 05 and 19
    RGI19              = 19
    RGI05 is dropped from emulandice (Ladrillo carries it in the Greenland target).
Ladrillo: julia/diag_gsic_blocks_vs_emulandice.jl, arm `joint` = posterior draw k on shared
sample k (the like-for-like arm), arm `fixed` = every draw on the mean shared path, arm
`joint_cmip6amp` = joint with each block's amplification overridden by the CMIP6
regional-characteristic ratio (the driver lever; table columns only).

BASELINE. Both series are re-referenced to 2005: FACTS by construction (baseyear 2005),
Ladrillo by subtracting its own 2005 value. A single-year rebase is acceptable here ONLY
because both are deterministic smooth model series on a smooth spliced driver (no
interannual noise to alias); it is not the convention for observational series.

PAIRING. Ladrillo sample k and emulandice sample k saw the same GSAT path, so the paired
difference (Ladrillo_k - emulandice_k) removes the climate spread; its median and 5-95%
are reported per block next to the unpaired medians.

Writes outputs/diag_gsic_blocks_vs_emulandice.csv and figures/diag_gsic_blocks_vs_emulandice.png.
    python3 python/diag_gsic_blocks_vs_emulandice.py [--scens=ssp126,ssp245,ssp585,vvVL,vvH]
"""
import os
import sys
import subprocess
import numpy as np
import pandas as pd
import netCDF4 as nc4
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

REPO = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
FACTS = os.path.expanduser("~/Documents/2026/CodeProjects/facts/experiments")
IN_DIR = os.path.join(REPO, "outputs", "diag_emu_blocks")
OUT_CSV = os.path.join(REPO, "outputs", "diag_gsic_blocks_vs_emulandice.csv")
OUT_PNG = os.path.join(REPO, "figures", "diag_gsic_blocks_vs_emulandice.png")
SCENS_DEFAULT = ["ssp126", "ssp245", "ssp585", "vvVL", "vvH"]
LABEL = {"ssp126": "SSP1-2.6", "ssp245": "SSP2-4.5", "ssp585": "SSP5-8.5",
         "vvVL": "vv Very Low", "vvH": "vv High"}
BLOCKS = {"SLOWG": [3, 9, 7, 6],
          "FASTG": [1, 2, 4, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18],
          "RGI19": [19]}
CODE_NAME = {"SLOWG": "SLOWP", "FASTG": "FAST", "RGI19": "R19"}   # names in the Julia output
BASE_YEAR = 2005
EMU_YEARS = list(range(2020, 2101, 10))
ARM = "joint"                     # the like-for-like arm; `fixed` is reported in the table only
EMU_LABEL = "FACTS emulandice (GlacierMIP2 GP, GSAT-driven)"
LAD_LABEL = "Ladrillo L24 (joint: draw k on shared sample k)"
TITLE = "Glacier melt since 2005 by block — Ladrillo L24 vs emulandice on the SAME 200 FaIR configs"


def commit():
    return subprocess.run(["git", "-C", REPO, "rev-parse", "--short", "HEAD"],
                          capture_output=True, text=True).stdout.strip() or "UNKNOWN"


def emulandice_blocks(scen):
    """{block: array (samples x EMU_YEARS)} in cm rel 2005, summed over the block's regions."""
    d = os.path.join(FACTS, f"global.shared.{scen}.n200", "output")
    out = {}
    for b, regs in BLOCKS.items():
        acc = None
        for r in regs:
            f = os.path.join(d, f"global.shared.{scen}.n200.emuglaciers.emulandice.glaciers_glac{r}_globalsl.nc")
            if not os.path.exists(f):
                raise SystemExit(f"missing {f}")
            ds = nc4.Dataset(f)
            yrs = [int(y) for y in ds.variables["years"][:]]
            assert yrs == EMU_YEARS, f"{f}: years {yrs}"
            assert int(ds.baseyear) == BASE_YEAR, f"{f}: baseyear {ds.baseyear}"
            x = np.asarray(ds.variables["sea_level_change"][:, :, 0], dtype=float) / 10.0   # mm -> cm
            acc = x if acc is None else acc + x
        out[b] = acc
    return out


def ladrillo_blocks(scen, arm):
    """{block: array (samples x EMU_YEARS)} in cm rel 2005 from the Julia dump."""
    f = os.path.join(IN_DIR, f"ladrillo_blocks_{scen}.csv")
    if not os.path.exists(f):
        raise SystemExit(f"missing {f} — run julia/diag_gsic_blocks_vs_emulandice.jl --scen={scen}")
    d = pd.read_csv(f)
    d = d[d.arm == arm]
    out = {}
    for b, code in CODE_NAME.items():
        w = d[d.block == code].pivot(index="sample", columns="year", values="melt_cm").sort_index()
        base = w[BASE_YEAR].values[:, None]
        out[b] = (w[EMU_YEARS].values - base)
    prov = open(f.replace('.csv', '_provenance.txt')).read().strip()
    return out, prov


def summarise(scens):
    rows = []
    fig, axes = plt.subplots(len(BLOCKS) + 1, len(scens), figsize=(3.6 * len(scens), 2.9 * (len(BLOCKS) + 1)),
                             sharex=True)
    prov = None
    for j, scen in enumerate(scens):
        emu = emulandice_blocks(scen)
        lad, prov = ladrillo_blocks(scen, ARM)
        ladf, _ = ladrillo_blocks(scen, "fixed")
        lada, _ = ladrillo_blocks(scen, "joint_cmip6amp")
        n = min(emu["SLOWG"].shape[0], lad["SLOWG"].shape[0])
        # [SAMPLE-COUNT] the pairing is by index; both sides must carry the same n
        assert emu["SLOWG"].shape[0] == lad["SLOWG"].shape[0] == n, (scen, emu["SLOWG"].shape, lad["SLOWG"].shape)
        panels = list(BLOCKS) + ["sum (excl. RGI 05)"]
        for i, b in enumerate(panels):
            if b in BLOCKS:
                e, l, lf, la = emu[b], lad[b], ladf[b], lada[b]
            else:
                e = sum(emu[k] for k in BLOCKS); l = sum(lad[k] for k in BLOCKS)
                lf = sum(ladf[k] for k in BLOCKS); la = sum(lada[k] for k in BLOCKS)
            i21 = EMU_YEARS.index(2100)
            diff = l[:, i21] - e[:, i21]
            rows.append(dict(scen=scen, block=b, year=2100, n=n,
                             emu_med=np.median(e[:, i21]), emu_p05=np.percentile(e[:, i21], 5), emu_p95=np.percentile(e[:, i21], 95),
                             lad_med=np.median(l[:, i21]), lad_p05=np.percentile(l[:, i21], 5), lad_p95=np.percentile(l[:, i21], 95),
                             lad_fixed_med=np.median(lf[:, i21]), lad_fixed_p05=np.percentile(lf[:, i21], 5), lad_fixed_p95=np.percentile(lf[:, i21], 95),
                             lad_cmip6amp_med=np.median(la[:, i21]), lad_cmip6amp_p05=np.percentile(la[:, i21], 5), lad_cmip6amp_p95=np.percentile(la[:, i21], 95),
                             ratio_cmip6amp=np.median(la[:, i21]) / np.median(e[:, i21]),
                             paired_diff_med=np.median(diff), paired_diff_p05=np.percentile(diff, 5), paired_diff_p95=np.percentile(diff, 95),
                             ratio_of_medians=np.median(l[:, i21]) / np.median(e[:, i21]),
                             frac_samples_lad_below=np.mean(diff < 0)))
            ax = axes[i, j]
            for arr, c, lab in ((e, "#1f77b4", EMU_LABEL), (l, "#d62728", LAD_LABEL)):
                ax.plot(EMU_YEARS, np.median(arr, 0), color=c, lw=2, label=lab)
                ax.fill_between(EMU_YEARS, np.percentile(arr, 5, 0), np.percentile(arr, 95, 0), color=c, alpha=0.15)
            if i == 0:
                ax.set_title(LABEL[scen])
            if j == 0:
                ax.set_ylabel(f"{b}\ncm since {BASE_YEAR}")
            ax.grid(alpha=0.3)
    axes[0, 0].legend(fontsize=7, loc="upper left")
    fig.suptitle(f"{TITLE}\nmedian and 5–95% over the {n} matched samples; emulandice regions summed to Ladrillo's blocks, RGI 05 dropped; commit {commit()}", fontsize=9)
    fig.tight_layout(rect=(0, 0, 1, 0.95))
    os.makedirs(os.path.dirname(OUT_PNG), exist_ok=True)
    fig.savefig(OUT_PNG, dpi=130)
    t = pd.DataFrame(rows)
    t["provenance"] = (f"{os.path.basename(__file__)} @ {commit()}; emulandice from FACTS global.shared.<scen>.n200 "
                       f"per-region glac*.nc (base {BASE_YEAR}, mm->cm); Ladrillo: {prov}")
    t.to_csv(OUT_CSV, index=False, float_format="%.4f")
    return t


if __name__ == "__main__":
    scens = [a.split("=", 1)[1].split(",") for a in sys.argv[1:] if a.startswith("--scens=")]
    scens = scens[0] if scens else SCENS_DEFAULT
    t = summarise(scens)
    print(f"\n2100 glacier melt since {BASE_YEAR} (cm), {ARM} arm vs emulandice, matched samples")
    print(f"{'scen':8s} {'block':18s} {'emu med [5-95]':22s} {'Ladrillo med [5-95]':22s} {'ratio':>6s} {'paired diff med [5-95]':>26s} {'frac<0':>7s} {'CMIP6-amp med':>14s} {'ratio':>6s}")
    for r in t.itertuples():
        print(f"{r.scen:8s} {r.block:18s} {r.emu_med:5.2f} [{r.emu_p05:5.2f},{r.emu_p95:5.2f}]     "
              f"{r.lad_med:5.2f} [{r.lad_p05:5.2f},{r.lad_p95:5.2f}]     {r.ratio_of_medians:6.2f} "
              f"{r.paired_diff_med:8.2f} [{r.paired_diff_p05:6.2f},{r.paired_diff_p95:6.2f}]   {r.frac_samples_lad_below:5.2f} "
              f"{r.lad_cmip6amp_med:14.2f} {r.ratio_cmip6amp:6.2f}")
    print(f"\nwrote {os.path.relpath(OUT_CSV, REPO)} and {os.path.relpath(OUT_PNG, REPO)}")
